transport/socks5/packet_listener_test.go

```diff
@@ -0,0 +1,107 @@
+package socks5
+
+import (
+	"bytes"
+	"context"
+	"net"
+	"testing"
+	"time"
+
+	"github.com/Jigsaw-Code/outline-sdk/transport"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+	"github.com/things-go/go-socks5"
+)
+
+func TestSOCKS5Associate(t *testing.T) {
+	// Create a local listener
+	// This creates a UDP server that responded to "ping"
+	// message with "pong" response
+	locIP := net.ParseIP("127.0.0.1")
+	// Create a local listener
+	echoServerAddr := &net.UDPAddr{IP: locIP, Port: 12199}
+	echoServer := setupUDPEchoServer(t, echoServerAddr)
+	defer echoServer.Close()
+
+	// Create a socks server to proxy "ping" message
+	cator := socks5.UserPassAuthenticator{Credentials: socks5.StaticCredentials{
+		"testusername": "testpassword",
+	}}
+	proxySrv := socks5.NewServer(
+		socks5.WithAuthMethods([]socks5.Authenticator{cator}),
+		//socks5.WithLogger(socks5.NewLogger(log.New(os.Stdout, "socks5: ", log.LstdFlags))),
+	)
+	// Start listening
+	proxyServerAddress := "127.0.0.1:12355"
+	go func() {
+		err := proxySrv.ListenAndServe("tcp", proxyServerAddress)
+		require.NoError(t, err)
+	}()
+	time.Sleep(10 * time.Millisecond)
```

Comment:
What is this sleep for? We can probably avoid it. It's usually bad to use sleep in tests, since variations in execution time can break the test.

Reply:
@fortuna This was there to ensure server is ready before continuing with tests. I agree if the server takes longer than 10 ms to start listening, the test is going to fail. I added an async method to wait until the server is ready to make this more robust. I also had this sleep in `stream_dialer_test.go` and replaced it with `waitUntilServerReady` there as well.

```go 
func waitUntilServerReady(addr string, ready chan<- bool) {
	for {
		conn, err := net.Dial("tcp", addr)
		if err == nil {
			conn.Close()
			ready <- true
			return
		}
		time.Sleep(10 * time.Millisecond)
	}
}
```

Reply:
@fortuna I missed an important detail. Yes you are right. I forgot I changed the server call from `ListenAndServe` to `Serve` and I begin to listen separately with the last change. My test result was related to previous implementation that combined Listen and Serve into one call. I removed all the waits and re-tested the code. I ran it around 50 times and it never failed without the delay.